astropy/coordinates/tests/test_separation.py

```diff
@@ -0,0 +1,346 @@
+        ),
+    ],
+    ids=lambda x: x.pytest_id,
+)
+@pytest.mark.parametrize("method", ["position_angle", "separation", "separation_3d"])
+def test_array_coords(coord_class, other_coord, method):
+    coord = coord_class(0 * u.deg, [0, 1, 2] * u.deg, [3, 5, 7] * u.pc)
+    assert_quantity_allclose(
+        getattr(coord, method)(other_coord.coord), getattr(other_coord, method)
+    )
+    assert_quantity_allclose(
+        getattr(other_coord.coord, method)(coord),
+        getattr(other_coord, f"reversed_{method}"),
+    )
+
+
+@pytest.mark.parametrize(
+    "coord",
+    [
+        pytest.param(FK5(1 * u.deg, 0 * u.deg, 1 * u.pc), id="FK5"),
+        pytest.param(
+            SkyCoord(1 * u.deg, 0 * u.deg, 1 * u.pc, frame="fk5"), id="SkyCoord"
+        ),
+    ],
+)
+@pytest.mark.parametrize(
+    "other_coord",
+    [
+        SeparationExpectation(
+            FK5(1 * u.deg, 0 * u.deg, 1 * u.pc, equinox="B1950"),
+            "FK5_B1950",
+            66.51310007 * u.deg,
+            0.69835342 * u.deg,
+            0.01218849 * u.pc,
+            246.50823798 * u.deg,
+        ),
+        SeparationExpectation(
+            SkyCoord(1 * u.deg, 0 * u.deg, 1 * u.pc, frame="fk5", equinox="B1950"),
+            "SkyCoord_B1950",
+            66.51310007 * u.deg,
+            0.69835342 * u.deg,
+            0.01218849 * u.pc,
+            246.50823798 * u.deg,
+        ),
+    ],
+    ids=lambda x: x.pytest_id,
+)
+@pytest.mark.parametrize("method", ["position_angle", "separation", "separation_3d"])
+def test_equinox_conversion(coord, other_coord, method):
+    """
+    Regression test for
+        - #868, #891: methods raised errors in case of frames with equinoxes
+        - #3106, #15659: unspecified equinox should be interpreted as frame default,
+            both in `SkyCoord` methods (#3106) and in `BaseCoordinateFrame` methods with
+            `SkyCoord` input (#15659).
+        - #5702: `position_angle()` didn't check if equinoxes differed
+    """
+    assert_quantity_allclose(
+        getattr(coord, method)(other_coord.coord), getattr(other_coord, method)
+    )
+    assert_quantity_allclose(
+        getattr(other_coord.coord, method)(coord),
+        getattr(other_coord, f"reversed_{method}"),
+    )
+
+
+@pytest.mark.parametrize("other_class", [SkyCoord, ICRS])
+@pytest.mark.parametrize("coord_class", [SkyCoord, ICRS])
+def test_separation_3d_dimensionless_distance(coord_class, other_class):
+    assert_quantity_allclose(
+        coord_class(35 * u.deg, 0 * u.deg, 3 * u.one).separation_3d(
+            other_class(125 * u.deg, 0 * u.deg, 4 * u.one)
+        ),
+        5 * u.one,
+    )
+
+
+@pytest.mark.parametrize("dimensionless_class", [SkyCoord, ICRS])
+@pytest.mark.parametrize("length_class", [SkyCoord, ICRS])
+def test_separation_3d_distance_dimension_mismatch(length_class, dimensionless_class):
+    dimensionless_coord = dimensionless_class(1 * u.deg, -2 * u.deg, 14)
```

Comment:
This feels weird - better not to put in any number, so one gets a `UnitSphericalRepresentation` - a distance without units should really lead to an error... 

EDIT: and now I see that case is tested below. I feel we should not test behaviour that is wrong, so would rather omit this test and have a separate PR that ensures we do not pass in dimensionless numbers for the distance...

Reply:
I didn't want to drop test cases and we do have explicit tests for this: https://github.com/astropy/astropy/blob/ea875472867f296eee3ed75989ed402d55587940/astropy/coordinates/tests/test_frames.py#L728-L732
But if you think this test really should be removed then I will.

Reply:
Arrgghh!  For this PR, though, more logical to keep the test indeed. It sounds like we'd need to deprecate dimensionless distance, which may not be worth it...